Use smarter type equality in `DependentTypeFinder`

brettfo opened the pull request:
_This is the same as #9056 but against `master`._

Due to potential assembly retargeting, simple type equivalence (e.g., `==`) can't be used in `DependentTypeFinder` so instead equivalence is checked via an added `AbstractReferenceFinder.OriginalSymbolsMatch` shortened stub.

All added tests force retargeting by referencing the _portable_ mscorlib in the base test assembly and the _normal_ mscorlib in the other.

The end-user impact is that when right-clicking on a type in the Solution Explorer and selecting "Derived Types" or "Implements", the complete (correct) set of types is shown.

Fixes #4973.

brettfo commented:
I also manually tested with all known issues to date:

Via #4973:

* `PreviewWorkspace` -> `PreviewDialogWorkspace`
* `IDocumentSupportsFeatureService`
* `AbstractSemanticQuickInfoSourceTests`

Via #5496:

* `AbstractGenerateMemberCodeFixProvider`

brettfo commented:
Tagging @dotnet/roslyn-ide for review.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs:
```diff
@@ -177,7 +178,7 @@ internal static class DependentTypeFinder
                     solution,
                     null,
-                    (t1, t2) => t1.BaseType == t2,
+                    (candidate, baseType) => AbstractReferenceFinder<INamedTypeSymbol>.OriginalSymbolsMatch(candidate.BaseType, baseType, solution, cancellationToken),
```
It feels like this helper should be moved elsewhere.

brettfo replied:
Do you mean rather than calling `AbstractReferenceFinder<>.OriginalSymbolsMatch()` this gets piped through `SomeOtherHelper.DoSymbolsMatch()` (which eventually calls `OriginalSymbolsMatch`)?

CyrusNajmabadi replied:
I mean: it's super weird for this static class to be calling another static method in a generic type elsewhere in order to get this functionality.   Either the static should be in DependentTypeFinder (and AbstractReferenceFinder should call that one instead).  Or it should be in some helper class that both of these finders use.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs:
```diff
@@ -244,6 +244,37 @@ protected static bool IdentifiersMatch(ISyntaxFactsService syntaxFacts, string n
+            }
+
+            Compilation searchSymbolCompilation = null;
+            if (!TryGetCompilation(searchSymbol, solution, ref searchSymbolCompilation, cancellationToken))
```
Why is this ref and not out?

brettfo replied:
Because `TryGetCompilation()` is using the `ref` as a cache.  The method looks like:

``` C#
if (compilation == null)
{
    compilation = //
}

return true;
```

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs:
```diff
@@ -244,6 +244,37 @@ protected static bool IdentifiersMatch(ISyntaxFactsService syntaxFacts, string n
+                return false;
+            }
+
+            return OriginalSymbolsMatch(searchSymbol, symbolToMatch, solution, ref searchSymbolCompilation, symbolToMatchCompilation, cancellationToken);
```
Why is this ref?

brettfo replied:
Short answer: I don't know.

Longer possible answer: eventually `OriginalSymbolsMatch()` calls `TryGetCompilation()` on `searchSymbolCompilation` which requires a `ref` (see earlier comment).  As to why only `searchSymbolCompilation` is `ref` and `symbolToMatchCompilation` isn't, ¯\\(°_o)/¯.

CyrusNajmabadi replied:
Can you find out? :)

Pilchie commented on src/Workspaces/CoreTest/DependentTypeFinderTests.cs:
```diff
@@ -0,0 +1,259 @@
+{
+    public abstract class BaseClass { }
+}
+", MscorlibRefPortable);
```
I'm not sure which mscorlib this actually is, but it might be worth testing against both a System.Runtime based portable and an "old-style" portable (like one that includes Silverlight).

brettfo replied:
Thanks, I'll confirm what type of portable this is and add a test for the other.

brettfo commented:
@CyrusNajmabadi: I've moved the helper like you suggested and got rid of the `ref` parameter (it was being used as a poor-man's cache which was able to be simplified by defaulting to passing in `null` and then if it ever needs to be resolved, the recursive calls will have it.)

@Pilchie: My initial tests were using an mscorlib-based portable library so I've added another test that uses a System.Runtime version.

Also, the mac failure is due to a crash in BoundTreeGenerator.exe so I'll also invoke `#notmyfault` (but I'll also re-run that with "test prtest/mac/dbg/unit32 please").

Pilchie commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs:
```diff
@@ -359,5 +359,257 @@ private static List<SymbolKey> GetOrAddDependentTypes(ConditionalWeakTable<Compi
+            out Compilation definingCompilation,
+            CancellationToken cancellationToken)
+        {
+            var definitionProject = solution.GetProject(symbol.ContainingAssembly, cancellationToken);
```
Should you be using `SymbolFinder.FindSourceDefinition` here first?

Pilchie commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs:
```diff
@@ -359,5 +359,257 @@ private static List<SymbolKey> GetOrAddDependentTypes(ConditionalWeakTable<Compi
+            }
+
+            // compilation from definition project must already exist.
+            if (!definitionProject.TryGetCompilation(out definingCompilation))
```
This can be null if no one has asked for it yet.  You're saying it should exist because you believe you have a symbol that you think should have come from it?

Pilchie replied:
Ah, now I see that all this code just moved.

Pilchie commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs:
```diff
@@ -359,5 +359,257 @@ private static List<SymbolKey> GetOrAddDependentTypes(ConditionalWeakTable<Compi
+
+            if (searchSymbol.Kind == SymbolKind.NamedType && symbolToMatch.IsConstructor())
+            {
+                return OriginalSymbolsMatch(searchSymbol, symbolToMatch.ContainingType, solution, searchSymbolCompilation, symbolToMatchCompilation, cancellationToken);
```
Wait constructors always match their containing type?  Seems fishy to me.

brettfo replied:
When wouldn't a constructor match it's containing type?

This code is a direct copypasta from the other file, but if it's incorrect then we might as well fix it now.

Pilchie commented:
:+1:

brettfo commented:
I've rebased to a current SHA and again manually verified the cases I called out in the first comment.

brettfo commented:
Closing in favor of #9056 (against `stabilization`).
